### PR TITLE
Add CheckboxGroup component

### PR DESCRIPTION
**Pull request by tranhoangan22**
Addresses issue [#1885](https://github.com/sumup-oss/circuit-ui/issues/1815)

## Purpose

Add the `CheckboxGroup` component.

## Approach and changes

- The UIs are implemented with inspirations from the `RadioButtonGroup` component.
- `CheckboxGroup` renders multiple Checkboxes by passing in the array `options` as prop which indicates how a `Checkbox` in the `CheckboxGroup` is rendered.
- Each option in the `options` prop must have at least a `value` key and a `label` key. The additional `label` key is the label for each individual `Checkbox`. 
- `CheckboxGroup` also accepts a `value` prop which is an array of the values of the Checkboxes within the `CheckboxGroup` that are checked by default.
- `validationHint` for the single `Checkbox` is [disallowed](https://github.com/sumup-oss/circuit-ui/issues/1815#issuecomment-1323647714) to make way for a single `validationHint` per `CheckboxGroup`.

## Definition of done

* [x] Development completed
* [x] Reviewers assigned
* [x] Unit and integration tests
* [ ] Meets minimum browser support
* [ ] Meets accessibility requirements

### CODE REVIEW

File: packages/circuit-ui/components/CheckboxGroup/CheckboxGroup.tsx [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+              style,
+              label: checkboxLabel,
+              ...rest
+            }) => (
+              <div
+                key={checkboxValue && checkboxValue.toString()}
+                className={className}
+                style={style}
+              >
+                <Checkbox
```
Comment (connor-baer) [inline packages/circuit-ui/components/CheckboxGroup/CheckboxGroup.tsx:186]: Why did you opt for a `div` around each `Checkbox` instead of passing the style props to the `Checkbox` directly?

File: packages/circuit-ui/components/CheckboxGroup/index.ts [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { CheckboxGroup } from './CheckboxGroup';
+
+export type { CheckboxGroupProps } from './CheckboxGroup';
+
+export default CheckboxGroup;
```
Comment (connor-baer) [inline packages/circuit-ui/components/CheckboxGroup/index.ts:20]: Don't forget to export the `CheckboxGroup` from the `packages/circuit-ui/index.ts` file as well :)

File: packages/circuit-ui/components/CheckboxGroup/CheckboxGroup.tsx (hunk 1 above)
Comment (tranhoangan22) [reply packages/circuit-ui/components/CheckboxGroup/CheckboxGroup.tsx:186]: I think we can indeed that we can pass the `key`, `className` and `style` props directly to the `Checkbox`. Just did it. 
I suppose we could do the same for the [`RadioButtonGroup`](https://github.com/sumup-oss/circuit-ui/blob/206b29f784f82e13f21133ac6d4961aac2fd21fe/packages/circuit-ui/components/RadioButtonGroup/RadioButtonGroup.tsx#L156-L160) right? @connor-baer @robinmetral

File: packages/circuit-ui/components/CheckboxGroup/index.ts (hunk 2 above)
Comment (tranhoangan22) [reply packages/circuit-ui/components/CheckboxGroup/index.ts:20]: Thanks for the pointer! Just did it :)

File: packages/circuit-ui/components/CheckboxGroup/CheckboxGroup.tsx (hunk 1 above)
Comment (connor-baer) [reply packages/circuit-ui/components/CheckboxGroup/CheckboxGroup.tsx]: > I suppose we could do the same for the [RadioButtonGroup](https://github.com/sumup-oss/circuit-ui/blob/206b29f784f82e13f21133ac6d4961aac2fd21fe/packages/circuit-ui/components/RadioButtonGroup/RadioButtonGroup.tsx#L156-L160) right?

Unlike the `Checkbox`, the `RadioButton` doesn't wrap its input and label in a wrapper `div`. I would add a wrapper to the `RadioButton` component in the next major to align the components (related to https://github.com/sumup-oss/circuit-ui/issues/1815#issuecomment-1323647714).

**Review by connor-baer — APPROVED** [review]
This looks great to me! Let's give @robinmetral a chance to take a final look when he's back tomorrow.

File: packages/circuit-ui/components/CheckboxGroup/CheckboxGroup.tsx (hunk 1 above)
Comment (robinmetral) [reply packages/circuit-ui/components/CheckboxGroup/CheckboxGroup.tsx]: Let's open an issue for this so we don't forget about it in v7 👀

Comment (connor-baer) [reply packages/circuit-ui/components/CheckboxGroup/CheckboxGroup.tsx]: Here ya go: https://github.com/sumup-oss/circuit-ui/issues/1875

**Review by robinmetral — COMMENTED** [review]
Thank you for the PR, this looks great already! ✨ And sorry for the late review.

I added a few thoughts below. Many of them are a minor suggestions for improvement, and a few tie back into how we want to component to function/be used. Let's see what we can address in here and if we need to have a chat about the broader questions 🙂 Let me know!

File: packages/circuit-ui/components/CheckboxGroup/CheckboxGroup.tsx [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+   */
+  showValid?: boolean;
+  /**
+   * Visually hide the label. This should only be used in rare cases and only
+   * if the purpose of the field can be inferred from other context.
+   */
+  hideLabel?: boolean;
+}
+
+const Legend = styled('legend')<Record<string, unknown>>(typography('two'));
```
[the remark this answers is not in the document]
Comment (tranhoangan22) [reply packages/circuit-ui/components/CheckboxGroup/CheckboxGroup.tsx:92]: I think that's because `Legend` wraps around the `FieldLabelText` so `Legend` expects a `prop` named `children`. I can rewrite it to be something like so.
```
const Legend = styled('legend')<Record<'children', JSX.Element>>(
  typography('two'),
);
```
WDYT?

File: packages/circuit-ui/components/CheckboxGroup/CheckboxGroup.tsx [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+          return accumulator;
+        },
+        {},
+      );
+    }
+
+    return (
+      <FieldWrapper
+        as="fieldset"
+        role="group"
```
[the remark this answers is not in the document]
Comment (tranhoangan22) [reply packages/circuit-ui/components/CheckboxGroup/CheckboxGroup.tsx:156]: Nice, didn't know this. Will remove this redundant prop!

File: packages/circuit-ui/components/CheckboxGroup/CheckboxGroup.tsx [hunk 5]
```diff
@@ -0,0 +1,214 @@
...
+        <Legend>
+          <FieldLabelText
+            label={label}
+            hideLabel={hideLabel}
+            optionalLabel={optionalLabel}
+            required={required}
+          />
+        </Legend>
+        {options &&
+          options.map(
```
[the remark this answers is not in the document]
Comment (tranhoangan22) [reply packages/circuit-ui/components/CheckboxGroup/CheckboxGroup.tsx:173]: Yes that's true. Removed it! :)

File: packages/circuit-ui/components/CheckboxGroup/CheckboxGroup.tsx [hunk 6]
```diff
@@ -0,0 +1,214 @@
...
+            ({
+              value: checkboxValue,
+              className,
+              style,
+              label: checkboxLabel,
+              ...rest
+            }) => (
+              <Checkbox
+                key={checkboxValue && checkboxValue.toString()}
```
[the remark this answers is not in the document]
Comment (tranhoangan22) [reply packages/circuit-ui/components/CheckboxGroup/CheckboxGroup.tsx:182]: I think the `name` prop is optional. Do you think it should be required? Nevertheless could do something like so:
> key={checkboxValue && `${checkboxValue.toString()}-${index}`}

File: packages/circuit-ui/components/CheckboxGroup/CheckboxGroup.tsx [hunk 7]
```diff
@@ -0,0 +1,214 @@
...
+                className={className}
+                style={style}
```
[the remark this answers is not in the document]
Comment (tranhoangan22) [reply packages/circuit-ui/components/CheckboxGroup/CheckboxGroup.tsx:184]: Yes, removed them. Thanks! :)

File: packages/circuit-ui/components/CheckboxGroup/CheckboxGroup.tsx [hunk 8]
```diff
@@ -0,0 +1,214 @@
...
+                {...{
+                  ...rest,
+                  value: checkboxValue,
+                  name,
+                  required,
+                  onChange,
+                  validationHint: undefined, // disallow `validationHint` for the single Checkbox
```
[the remark this answers is not in the document]
Comment (tranhoangan22) [reply packages/circuit-ui/components/CheckboxGroup/CheckboxGroup.tsx:191]: Done :)

File: packages/circuit-ui/components/CheckboxGroup/CheckboxGroup.tsx (hunk 5 above)
[the remark this answers is not in the document]
Comment (tranhoangan22) [reply packages/circuit-ui/components/CheckboxGroup/CheckboxGroup.tsx:173]: Very nice point! I've wrapped the checkbox group around with an `ul` and the individual checkboxes with `li`s :)

File: packages/circuit-ui/components/CheckboxGroup/CheckboxGroup.stories.tsx [hunk 9]
```diff
@@ -0,0 +1,238 @@
...
+      const index = value.indexOf(onChangeValue);
+      if (index === -1) {
+        setValue([...value, onChangeValue]);
+      } else {
+        setValue([...value.slice(0, index), ...value.slice(index + 1)]);
+      }
+    } else {
+      setValue([event.target.value]);
+    }
+  };
```
[the remark this answers is not in the document]
Comment (tranhoangan22) [reply packages/circuit-ui/components/CheckboxGroup/CheckboxGroup.stories.tsx:49]: That is true. The implementation is indeed cumbersome. Perhaps there are ways to make life easier for the users of this components in the controlled cases! I gave it some thought before opting for the the `value` being an array of individual checkboxes' values. Perhaps there are better ways. Let's talk about it!

File: packages/circuit-ui/components/CheckboxGroup/CheckboxGroup.tsx [hunk 10]
```diff
@@ -0,0 +1,214 @@
...
+   */
+  optionalLabel?: string;
+  /**
+   * The ref to the HTML DOM element.
+   */
+  ref?: Ref<HTMLFieldSetElement>;
+  /**
+   * Makes the input group required.
+   */
+  required?: InputHTMLAttributes<HTMLInputElement>['required'];
```
[the remark this answers is not in the document]
Comment (tranhoangan22) [reply packages/circuit-ui/components/CheckboxGroup/CheckboxGroup.tsx:68]: Yes that's a very valid point! It's better to be able to specify which checkboxes are required. I've included the `required` prop for each `option`, and removed the `optionalLabel`. 
 I guess we wouldn't need a `optionalLabel` for an optional checkbox (i.e., `required` is falsy)? Or do you think we should have a label for the group in case some of the checkboxes are optional?

File: packages/circuit-ui/components/CheckboxGroup/CheckboxGroup.tsx [hunk 11]
```diff
@@ -0,0 +1,214 @@
...
+      process.env.NODE_ENV !== 'production' &&
+      process.env.NODE_ENV !== 'test' &&
+      !label
+    ) {
+      throw new AccessibilityError(
+        'CheckboxGroup',
+        'The `label` prop is missing. Pass `hideLabel` if you intend to hide the label visually.',
+      );
+    }
+    const name = customName || uniqueId('checkbox-button-group_');
```
[the remark this answers is not in the document]
Comment (tranhoangan22) [reply packages/circuit-ui/components/CheckboxGroup/CheckboxGroup.tsx:134]: Thanks for spotting this! We can have a "custom" `name` (i.e., `checkbox-button-group_`${name}``) for the whole checkbox group and allow developers to pass the names for the individual checkboxes via the `options` prop. I don't think it's correct to have all the checkboxes in the checkbox group having the same name. I took inspirations from the implementation of the `RadioButtonGroup` where the individual buttons should share the same `name` as the group itself (which has a custom name) for the radio button group to work correctly. This is obviously not the case for the CheckboxGroup.
I've stopped passing the checkbox group's customName to the individual checkboxes. Let me know if you think this could be implemented better!

File: packages/circuit-ui/components/CheckboxGroup/CheckboxGroup.tsx (hunk 10 above)
[the remark this answers is not in the document]
Comment (connor-baer) [reply packages/circuit-ui/components/CheckboxGroup/CheckboxGroup.tsx]: I can think of two scenarios of required checkboxes in a group:

1. At least _x_ amount of options need to be checked. Unfortunately, [HTML5 does not directly support](https://www.w3.org/html/wg/tracker/issues/111?changelog) requiring only one/at least one checkbox to be checked in a checkbox group, so this needs to be validated with JavaScript by the developers.
2. A specific subset of options needs to be checked. This can be achieved by passing the `required` attribute to these specific options. However, this seems like an anti-pattern: it would be more user-friendly to precheck the options and disable them to indicate to the users that these options cannot/should not be changed.

In other words, I don't think the `required` is useful in the context of a checkbox group.

**Review by connor-baer — COMMENTED** [review]
Right now, there are two ways the state of the checkboxes can be controlled: the `checked`/`defaultChecked` props on the individual options or the `value`(/`defaultValue`?) prop on the group. 

Should we allow both approaches, which could get complicated if developers mix the two? If so, we should clarify in the documentation which takes precedence.

**Comment by tranhoangan22** [comment]
> Should we allow both approaches, which could get complicated if developers mix the two? If so, we should clarify in the documentation which takes precedence

I think we should avoid complication and confusion by only allowing passing the `value` prop on the group. The current implementation already makes sure this takes precedence (over the `checked` key in the `options` for each checkbox). We're lacking documentation for it though; I will add it. 

The use of the `defaultValue` prop on the group has not been addressed as a precedence over that of the `defaultChecked` on the individual checkbox. Do you think this is necessary?  

I'm wondering about the need to pass in the `defaultValue` prop (and `defaultChecked` in `options` for an individual checkbox) in the case of an uncontrolled CheckboxGroup. I think we can also have uncontrolled CheckboxGroups with the `value` prop (by not keeping track of the `value` via a piece of state when we implement it) ? 🤔

**Comment by connor-baer** [comment]
> I think we should avoid complication and confusion by only allowing passing the `value` prop on the group. The current implementation already makes sure this takes precedence (over the `checked` key in the `options` for each checkbox). We're lacking documentation for it though; I will add it.

👌

> The use of the `defaultValue` prop on the group has not been addressed as a precedence over that of the `defaultChecked` on the individual checkbox. Do you think this is necessary?
>
> I'm wondering about the need to pass in the `defaultValue` prop (and `defaultChecked` in `options` for an individual checkbox) in the case of an uncontrolled CheckboxGroup. I think we can also have uncontrolled CheckboxGroups with the `value` prop (by not keeping track of the `value` via a piece of state when we implement it)? 🤔

`value` and `defaultValue` are treated differently by React. React uses `defaultValue` to set the _initial_ value of an input and then stops tracking it. React uses `value` to always force the value of the input and thus forces it to be controlled. If you didn't pass an `onChange` handler, the value would never change (the [React docs](https://react.dev/reference/react-dom/components/input#controlling-an-input-with-a-state-variable) have an example).

So yes, we should add support for the `defaultValue` prop. If we disallow the `checked` prop on the options, we should also disallow the `defaultChecked` prop.

File: packages/circuit-ui/components/CheckboxGroup/CheckboxGroup.tsx (hunk 10 above)
[the remark this answers is not in the document]
Comment (connor-baer) [reply packages/circuit-ui/components/CheckboxGroup/CheckboxGroup.tsx]: I've changed my mind: we should allow the `required` prop, not for semantic reasons, but to toggle the `optionalLabel`.

File: packages/circuit-ui/components/CheckboxGroup/CheckboxGroup.tsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (connor-baer) [reply packages/circuit-ui/components/CheckboxGroup/CheckboxGroup.tsx]: Why does the generic type need to be provided at all? If TypeScript otherwise complains about the missing `theme` prop, you can use the `NoTheme` helper from `styles/styled`.

File: packages/circuit-ui/components/CheckboxGroup/CheckboxGroup.tsx [hunk 12]
```diff
@@ -0,0 +1,214 @@
...
+        role="group"
+        aria-describedby={descriptionIds}
+        name={name}
+        // @ts-expect-error TypeScript isn't smart enough to recognize the `as` prop.
+        ref={ref}
+        disabled={disabled}
+        {...props}
+      >
+        <Legend>
+          <FieldLabelText
```
[the remark this answers is not in the document]
Comment (connor-baer) [reply packages/circuit-ui/components/CheckboxGroup/CheckboxGroup.tsx]: The `FieldLabelText` does more than apply some styles, it's also responsible for the conditional `optionalLabel`.

File: packages/circuit-ui/components/CheckboxGroup/CheckboxGroup.tsx (hunk 6 above)
[the remark this answers is not in the document]
Comment (connor-baer) [reply packages/circuit-ui/components/CheckboxGroup/CheckboxGroup.tsx]: The `name` prop should be consistent for all options. We've switched to the option `label` for the `key`.

File: packages/circuit-ui/components/CheckboxGroup/CheckboxGroup.stories.tsx (hunk 9 above)
[the remark this answers is not in the document]
Comment (connor-baer) [reply packages/circuit-ui/components/CheckboxGroup/CheckboxGroup.stories.tsx]: @tranhoangan22 and I decided to stick with the array shape to match the SelectorGroup component. We were able to refactor the state update logic to simplify it. We'll document it in Storybook.

**Review by connor-baer — APPROVED** [review]
This is fantastic! Thank you for the contribution 🙌

In a follow-up PR, I'll add documentation and a few tweaks to align it with the other boolean inputs.
